pkg/resources/webhook/service.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2022 The Kubermatic Kubernetes Platform contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"fmt"
+
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/apis/kubermatic/v1"
+	"k8c.io/kubermatic/v2/pkg/resources"
+	"k8c.io/kubermatic/v2/pkg/resources/certificates/triple"
+	"k8c.io/kubermatic/v2/pkg/resources/reconciling"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	certutil "k8s.io/client-go/util/cert"
+)
+
+// ServiceCreator returns the function to reconcile the usercluster webhook service.
+func ServiceCreator() reconciling.NamedServiceCreatorGetter {
+	return func() (string, reconciling.ServiceCreator) {
+		return resources.UserClusterWebhookServiceName, func(se *corev1.Service) (*corev1.Service, error) {
+			se.Name = resources.UserClusterWebhookServiceName
+			se.Labels = resources.BaseAppLabels(resources.UserClusterWebhookServiceName, nil)
+
+			se.Spec.Type = corev1.ServiceTypeClusterIP
+			se.Spec.Selector = map[string]string{
+				resources.AppLabelKey: resources.UserClusterWebhookDeploymentName,
+			}
+			se.Spec.Ports = []corev1.ServicePort{
+				{
+					Name:       "",
+					Port:       443,
+					Protocol:   corev1.ProtocolTCP,
+					TargetPort: intstr.FromInt(9443),
+				},
+			}
+
+			return se, nil
+		}
+	}
+}
+
+type tlsServingCertCreatorData interface {
+	GetRootCA() (*triple.KeyPair, error)
+	Cluster() *kubermaticv1.Cluster
+}
+
+// TLSServingCertificateCreator returns a function to create/update the secret with the machine-webhook tls certificate.
+func TLSServingCertificateCreator(data tlsServingCertCreatorData) reconciling.NamedSecretCreatorGetter {
+	return func() (string, reconciling.SecretCreator) {
+		return resources.MachineWebhookServingCertSecretName, func(se *corev1.Secret) (*corev1.Secret, error) {
```

Comment:
`resources.MachineWebhookServingCertSecretName` We should have a dedicated secret for this.